lib/model/model.go

```diff
@@ -2550,24 +2556,39 @@ func makeForgetUpdate(files []protocol.FileInfo) []protocol.FileDownloadProgress
 
 // shouldIgnore returns true when a file should be excluded from processing
 func shouldIgnore(file db.FileIntf, matcher *ignore.Matcher, ignoreDelete bool) bool {
-	// We check things in a certain order here...
-
 	switch {
 	case ignoreDelete && file.IsDeleted():
 		// ignoreDelete first because it's a very cheap test so a win if it
 		// succeeds, and we might in the long run accumulate quite a few
 		// deleted files.
 		return true
 
+	case isInternal(file.FileName()):
+		return true
+
 	case matcher.Match(file.FileName()).IsIgnored():
-		// ignore patterns second because ignoring them is a valid way to
-		// silence warnings about them being invalid and so on.
 		return true
 	}
 
 	return false
 }
 
+func isInternal(name string) bool {
+	for _, internal := range []string{".stfolder", ".stignore", ".stversions"} {
+		// The internal always-ignored file names should never be
+		// transferred. It is however valid for them to appear in some
+		// deeper place in the hierarchy (i.e. someOtherFolder/.stignore is
+		// just a file as far as we're concerned.)
+		if name == internal {
+			return true
+		}
+		if strings.HasPrefix(name, internal+string(os.PathSeparator)) {
+			return true
+		}
+	}
+	return false
```

Comment:
Temp files I'm not convinced about belongs in the same place, but yes there should be a central place where the knowledge about what files are internal lives. This is a good as any. But then it needs to live somewhere else - `config` perhaps, or `scanner`. Hmm.